addons/mail/static/src/components/discuss/tests/discuss_tests.js

```diff
@@ -4206,6 +4206,125 @@ QUnit.test('auto-focus composer on opening thread', async function (assert) {
     );
 });
 
+QUnit.test('mark channel as seen if last message visible when switching channels [REQUIRE FOCUS]', async function (assert) {
+    assert.expect(6);
+
+    // channel expected to be found in the sidebar, with the expected message_unread_counter
+    // and a random unique id that will be referenced in the test
+    this.data['mail.channel'].records.push(
+        { id: 10, message_unread_counter: 1, name: 'Bla' },
+        { id: 11, message_unread_counter: 1, name: 'Blu' },
+    );
+    this.data['mail.message'].records.push({
+        body: 'oldest message',
+        channel_ids: [10],
+        id: 10,
+    }, {
+        body: 'newest message',
+        channel_ids: [11],
+        id: 11,
+    });
+    await this.start();
+    assert.containsOnce(
+        document.body,
+        `.o_DiscussSidebar_item[data-thread-local-id="${
+            this.env.models['mail.thread'].findFromIdentifyingData({
+                id: 10,
+                model: 'mail.channel',
+            }).localId
+        }"]`,
+        "should have discuss sidebar item with the channel #10"
+    );
+    assert.containsOnce(
+        document.body,
+        `.o_DiscussSidebar_item[data-thread-local-id="${
+            this.env.models['mail.thread'].findFromIdentifyingData({
+                id: 11,
+                model: 'mail.channel',
+            }).localId
+        }"]`,
+        "should have discuss sidebar item with the channel #11"
+    );
+    assert.hasClass(
+        document.querySelector(`
+            .o_DiscussSidebar_item[data-thread-local-id="${
+                this.env.models['mail.thread'].findFromIdentifyingData({
+                    id: 10,
+                    model: 'mail.channel',
+                }).localId
+            }"]
+        `),
+        'o-unread',
+        "sidebar item of channel #10 should be unread"
+    );
+    assert.hasClass(
+        document.querySelector(`
+            .o_DiscussSidebar_item[data-thread-local-id="${
+                this.env.models['mail.thread'].findFromIdentifyingData({
+                    id: 11,
+                    model: 'mail.channel',
+                }).localId
+            }"]
+        `),
+        'o-unread',
+        "sidebar item of channel #11 should be unread"
+    );
```

Comment:
Not sure whether it's covered by other(s) test(s), but should be dedicated test.

Reply:
I don't agree, the intent of the test is to check that the switching make the channel becomes read so it needs to check it's unread before doing the action. Otherwise the test is not robust enough if something else before the action made it unread.

Reply:
Problem with having assertions before is that test is less maintainable overall. It feels right but it isn't, that an illusion. You think that making test crash if premises aren't respected is good, but in the end it adds more noise in QUnit test crash and also makes test much longer than it should.

The purpose of **_this_** test is to check that is marks thread as read on switch when last message is visible. It's fine if it passes due to thread being already marked as read initially: another test should cover that exact scenario.

Reply:
Still not totally convinced, but fine

Reply:
Do we have this other test right now, and if so which one is it?

I feel like if we really fully want to go this way, we NEED to be able to express test dependencies, and if only manually with comments.

Reply:
> I feel like if we really fully want to go this way, we NEED to be able to express test dependencies, even if only manually with comments.

Really annoying QUnit lacking way to express dependencies among tests... To the point I would have liked not using QUnit at all and just have our own tools for unit testing...

Expressing dependencies in comment, considering how weak it is (not visible in QUnit UI, debug requires reading actual comment, likely comment becomes outdated, etc.), I prefer providing expected working behaviour in plain English rather than referencing other test. I know it sucks, but that's the best I can find without actual support of test dependencies...

Reply:
Well using phrases to describe initial state is more complicated and less robust than just keeping initial assertions I think...

Only dependencies avoids duplicating assertion/explanation.

Reply:
Honestly can't find decent replacement for lack of dependencies... Putting assertions in premise was our choice since not long ago, but these unit tests become hard to maintain, even when they are nicely written.

Of course they are still much better than JS tests outside of discuss, it's night and day... but the increase of maintenance cost of this way of writing test is too significant for my taste. This problem is crescendo, so better improve now than later.

Reply:
I agree for new solution, and I think it's acceptable and better than alternatives too (I was never in favor of copy/pasting all those assertions in the first place anyway), just have to keep in mind we want dependencies or at least ability to execute tests in certain order.

Considering everything we are writing ourselves at Odoo, rewriting a test framework for that purpose doesn't even look like a big task to be honest.

Reply:
anything to change here ?

Reply:
let's consider it fine for now, just to keep in mind